lang/udf_cache.hh

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class udf_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    struct cache_entry_type {
+        mutable udf_instance inst;
```

Comment:
Why mutable?

Reply:
It's needed if we want to std::move from the loading cache, maybe a shared ptr would be also good
It won't be needed when we move away from the loading cache